crates/katana/core/src/backend/gas_oracle.rs

```diff
@@ -1,25 +1,326 @@
-        self.gas_prices.clone()
+        match self {
+            L1GasOracle::Fixed(fixed) => fixed.current_gas_prices(),
+            L1GasOracle::Sampled(sampled) => sampled
+                .prices
+                .try_lock()
+                .map(|prices| prices.gas_prices.clone())
+                .unwrap_or_else(|_| GasPrices::default()),
+        }
     }
 
     /// Returns the current data gas prices.
+    pub fn current_data_gas_prices(&self) -> GasPrices {
+        match self {
+            L1GasOracle::Fixed(fixed) => fixed.current_data_gas_prices(),
+            L1GasOracle::Sampled(sampled) => sampled
+                .prices
+                .try_lock()
+                .map(|prices| prices.data_gas_prices.clone())
+                .unwrap_or_else(|_| GasPrices::default()),
+        }
+    }
+
+    pub fn run_worker(&self, task_spawner: TaskSpawner) {
+        match self {
+            Self::Fixed(..) => {}
+            Self::Sampled(oracle) => {
+                let mut worker =
+                    GasOracleWorker::new(oracle.prices.clone(), oracle.l1_provider.clone());
+                task_spawner
+                    .build_task()
+                    .graceful_shutdown()
+                    .name("L1 Gas Oracle worker")
+                    .spawn(async move { worker.run().await });
+            }
+        }
+    }
+}
+
+impl SampledL1GasOracle {
+    pub async fn current_data_gas_prices(&self) -> GasPrices {
+        self.prices.lock().await.data_gas_prices.clone()
+    }
+
+    pub async fn current_gas_prices(&self) -> GasPrices {
+        self.prices.lock().await.gas_prices.clone()
+    }
+}
+
+impl FixedL1GasOracle {
     pub fn current_data_gas_prices(&self) -> GasPrices {
         self.data_gas_prices.clone()
     }
+
+    pub fn current_gas_prices(&self) -> GasPrices {
+        self.gas_prices.clone()
+    }
+}
+
+async fn update_gas_price<P: Provider<T>, T: Transport + Clone>(
+    l1_oracle: &mut SampledPrices,
+    provider: P,
+    buffer: &mut GasPriceBuffer,
+) -> anyhow::Result<()> {
+    // Attempt to get the gas price from L1
+    let last_block_number = provider.get_block_number().await?;
+
+    let fee_history =
+        provider.get_fee_history(1, BlockNumberOrTag::Number(last_block_number), &[]).await?;
+
+    let latest_gas_price = fee_history.base_fee_per_gas.last().context("Getting eth gas price")?;
+
+    buffer.add_sample(*latest_gas_price);
+
+    let blob_fee_history = fee_history.base_fee_per_blob_gas;
+    let avg_blob_base_fee = blob_fee_history.iter().last().context("Getting blob gas price")?;
+
+    let avg_blob_fee_eth = *avg_blob_base_fee;
+    let avg_blob_fee_strk = *avg_blob_base_fee + ONE_GWEI;
+
+    let avg_gas_price = GasPrices { eth: buffer.average(), strk: buffer.average() + ONE_GWEI };
+    let avg_blob_price = GasPrices { eth: avg_blob_fee_eth, strk: avg_blob_fee_strk };
+
+    l1_oracle.gas_prices = avg_gas_price;
+    l1_oracle.data_gas_prices = avg_blob_price;
+    Ok(())
```

Comment:
why don't you compute the average for the blob/data price as well?

Reply:
I overlooked this, I'll make sure to have it fixed